Reduce disambiguation tokens, instantiate function values. (#1823)
dart-lang/language: accepted/future-releases/constructor-tearoffs/feature-specification.md

  * Function tear-offs (Example: `var intId = id<int>;`)
* Implicit and explicit instantiation of function values (not just at tear-off).
  * `var id = <T>(T value) => value; int Function(int) intId = id;` (implicit).
  * `var intId = id<int>;` (explicit).

The new syntax causes some ambiguities in the grammar, which we resolve by a one-token lookahead after a potential type argument list.

### Named constructor tear-off

We allow tearing off named constructors.

If *C* denotes a class declaration (it's an identifier or qualified identifier which resolves to the class declaration) and *C.name* is the name of a constructor of that class, we allow you to tear off that constructors as:

* <code>*C*.*name*</code>, or
* <code>*C*\<*typeArgs*>.*name*</code>

just as you can currently invoke the constructor as <code>*C*.*name*(*args*)</code>, or <code>*C*\<*typeArgs*>.*name*(*args*)</code>.

Expressions of the form <code>*C*\<*typeArgs*>.*name*</code> are potentially compile-time constant expressions and are compile-time constants if the type arguments are constant types (and <code>*C*.*name*</code> actually denotes a constructor).

> _The former syntax, without type arguments, is currently allowed by the language grammar, but is rejected by the static semantics as not being a valid expression when denoting a constructor. The latter syntax is not currently grammatically an_ expression_. Both can occur as part of a constructor invocation, but cannot be expressions by themselves because they have no values. We introduce a static and dynamic expression semantic for such a *named constructor tear-off expression*, which makes them valid expressions._

Author: Done.

The static type of the named constructor tear-off expression is the same as the static type of the corresponding constructor function tear-off.

_We effectively desugar constructor tear-off into static method tear-off of a corresponding static function, which ensures that the behavior of constructor tear-offs is consistent with the behavior of static method tear-offs._

The new syntactic options introduce an **ambiguity** in the grammar. If `List.filled` is a valid expression, then `List.filled(4, 4)` can both be a constructor invocation *and* a tear-off followed by a function invocation, and `List.filled<int>(4, 4)` can *only* be valid as a tear-off followed by a function invocation.
This is similar to the existing possible ambiguity for an instance method invocation like `o.m(arg)`, and we resolve it the same way, by always preferring the direct invocation over doing a tear-off and a function invocation.

> We do not allow `List.filled<int>(4, 4)` at all.  We could allow it, it's syntactically similar to a getter invocation like `o.getter<int>(4)`, which we do handle without issues, but allowing that syntax to be a constructor tear-off could interfere with a possible later introduction of *generic constructors*. We only do constructor tear-off when the constructor reference is *not* followed by a *typeArguments* or *arguments* production. If it is followed by those, then it's a constructor *invocation*, and it's currently an error if such one includes type arguments. That is, an expression of the form `List.filled<int>(4, 4)` is *invalid*, it's not interpreted as a constructor tear-off followed by a function invocation, but always as a constructor invocation&mdash;and the `List.filled` constructor is not a generic construct (no constructor is, yet). You can write `(List.filled)<int>(4, 4)` to do the tear-off or `List<int>.filled(4, 4)` to do the invocation.

Author: Reworded.

Example:

```dart
var makeIntList = NumList<int>.filled; // Equivalent to `List<int>.filled` or `List.filled$tearoff<int>`List<double> Function(int, double) makeDoubleList = NumList.filled; // Same as `List<double>.filled` after inference.
```

Here `NumList<int>` is a single type (`List<int>`), and the tear-off happens from that type.

Notice that whether an alias expansion is constant depends on the parameters, not the result. Example:

```dart

> typedef Ignore2<T, S> = List<T>;void foo<X>() {  var c = Ignore2<int, X>.filled; // Aka List<int>.filled, but is *not constant*.}

Author: Ack, thought I caught them all. (Did a cut-n-repaste of the entire document to fix a quirk in my editor. Apparently that wasn't safe.)

```dart
typedef Ignore2<T, S> = List<T>;
void foo<X>() {

>   var c = Ignore2<int, X>.filled; // Aka. List<int>.filled, but is *not constant*.

Author: ACK. Our treatment of *constants* is generally very syntactical. We require the expression to be *potentially constant*, which means having one of a set of known syntactic forms, and then to have parts which are constant, which would exclude the `X` above. 

I'd prefer to keep this for now.

```dart
typedef ListList<T> = List<List<T>>;
// Corresponding factory 

> functionList<List<T>> ListList$filled$tearoff<T>(int length, List<T> value) => 

Author: It should.

```dart
var f = MyList.filled; // Equivalent to `List.filled` or `List.filled$tearoff`
// Instantiated type aliases use the aliased type, constant and canonicalized when type are constant.
print(identical(MyList<int>.filled, NumList<int>.filled)); // true
print(identical(MyList<int>.filled, List<int>.filled)); // true

> print(identical(NumList<int>.filled, List<int>.filled)); // true  

Author: Probably same issue as the other removal of newlines.

  C.name();
  List<T> createList() => <T>[];
}
typedef A<T extends dynamic> = C<T>; 

> // Proper rename, different, but equivalent, bound.

Author: Nope, mistake. Should either be on the same line or the previous line.
This was a mistakenly reinserted newline.

void main() {
  // Static type : C<T> Function<T extends Object?>()  
  // Runtime type: C<T> Function<T extends Object?>()
  var cf = C.name;
  // Static type : C<T> Function<T extends dynamic>()
  // Runtime type: C<T> Function<T extends Object?>()
  var af = A.name;
  var co = (cf as dynamic)();

>   var ao = (af as dynamic)(); 

Author: Here the comment is short enough to be on the same line.

>   var ao = (af as dynamic)(); 

Reviewer: It looks like this hasn't been fixed yet

Because of that, we introduce a *new* syntax that can be used to denote the unnamed constructor: <code>*C*.new</code>. It can be used in every place where a named constructor can be referenced, but will instead denote the unnamed constructor, *and* it can be used to tear off the unnamed constructor without interfering with using the class name to denote the `Type` object.

```dart

> class C {  

Author: Nope, also accidental.
I'm slightly surprised there aren't *more*, I had to reflow the entire code block.

*With regard to tear-offs, <code>C.new</code> works exactly as if it had been a named constructor, with a corresponding constructor function named <code>C.new$tearoff</code>.*

We probably want to support `[C.new]` as a constructor link in DartDoc as well. In `dart:mirrors`, the name of the constructor is still just `C`, not `C.new` (we don't want to break existing reflection using code).

> The grammar will be changed to allow ``<identifier> |`new'`` anywhere we currently denote a named constructor name, and we make it a *primary* expression to tear-off an unnamed constructor as `classRef.new`.

Author: Done.

```dart
T id<T>(T value) => value;
int Function(int) idInt = id; // Implicitly instantiated tear-off.
// and
typedef IntList = List<int>;
Type intList = IntList;
```

This feature introduces syntax to explicitly instantiate a generic function object or tear-off and a type literal for a generic type. The former for consistency with constructor tear-offs, the latter to introduce in-line types without needing a `typedef`, like we did for in-line function types originally. We introduce both now because they share the same grammar productions.

Example:

```dart
T id<T>(T value) => value;
var idInt = id<int>; // Explicitly instantiated tear-off, saves on function types.
// and
Type intList = List<int>; // In-line instantiated type literal.
```

> These grammar changes allow *type parameters* without following parenthesized arguments in places where we previously did not allow such. This means that `<typeArguments>` becomes a *selector* by itself, not just when followed by arguments.

Author: How about "did not allow that." 
The "them" can refer to both "type parameters" or "following parenthesized arguments" (both plural nouns earlier in the sentence). It's more obvious that it's type parameters, because of the recurrence of "allow", but it's not unambiguous. I want to refer to allowing the combination of "type parameters without a following argument list".

> These grammar changes allow *type parameters* without following parenthesized arguments in places where we previously did not allow such. This means that `<typeArguments>` becomes a *selector* by itself, not just when followed by arguments.

Reviewer: Sure, "did not allow that" works for me.

Explicit instantiation applies to any function value, whether tear-offs of instance methods or local, static and top-level function declarations, or first-class function values.

For an expression of the form <code>*e*\<*typeArgs*></code>, which is not followed by an argument list (that would turn it into a generic function invocation), then:

> * If *e* denotes a generic class, mixin or type alias, then <code>*e*\<*typeArgs*></code> is a type literal. _If followed by <code>.*id*</code> then that *id* must denote a constructor, which can then be either torn off or invoked. If followed by `==` or `!=` or any "stop-token", the expression evaluates to a `Type` object._

Author: If "*e* denotes a generic class, mixin or type alias" then it is necessarily of the form `id` or `prefix.id`.
The expression `(ClassName)` does not denote a class, it's just an expression which evaluates to a `Type` object.

I think I wrote that somewhere else (but could have been lost in editing). I'll put it here too.

> * If *e* denotes a top-level, static or local function declaration, it must be a generic function declaration, and then <code>*e*\<*typeArgs*></code> performs an explicitly instantiated function tear-off, which works just like the current implicitly instantiated function tear-off except that the types are provided instead of inferred.

Author: Yes, if *e* denotes a local function, it's necessarily an identifier. If it denotes a top-level function, it's necessarily an identifier or prefix.identifier. If it denotes a static method, it's one of identifier, *ClassOrMixinDeclaration*.identifier, or *prefix*.*ClassOrMixinDeclaration*.identifier.

  * It also includes *e* denoting a constructor. _(We reserve this syntax for denoting instantiation of generic constructors, should the language add [generic constructors](https://github.com/dart-lang/language/issues/647) in the future. Instead just write <code>(*C*.*name*)\<*typeArgs*\></code> or <code>*C*\<typeArgs\>.*name*</code>.)_

Cascades can contain explicitly instantiated tearoffs, because they can contain any selector and instantiation is now a selector, e.g., `receiver..foo()..instanceMethod<int>..bar`. _Note that this example is allowed for consistency, but it will compute a value and discard it. Instantiation without immediate invocation is expected to be primarily used in places where the value of that instantiation will be stored for later use, and using it in a cascade is outside of that usage pattern. One example where it could be useful would be as a receiver for an extension method on function types, like  `receiver..foo()..bar<int>.apply(argList)`. The first selector of a cascade section must still be one of `..identifier` or `..[index]`, it cannot be `..<typeArgs>` any more than it can be `..(argumentList)`._

```dart

> class A {  

Author: Blanks restored.

  List<X> m<X>(X x) => [x];
}
extension FunctionApplier on Function {
  void applyAndPrint(List<Object?> positionalArguments) =>
      print(Function.apply(this, positionalArguments, const {}));
}
void main() {  
  A()
    ..m<int>.applyAndPrint([2])
    ..m<String>.applyAndPrint(['three']);
}
```

The static type of the explicitly instantiated tear-offs are the same as if the type parameter had been inferred, but no longer depends on the context type. Missing type arguments in implicit instantiation expressions can now be considered "filled in" by type inference, as if they had been written explicitly, just as for other inferred type arguments.

The static type of the instantiated type literal is `Type`. This feature also satisfies issue [#123](https://github.com/dart-lang/language/issues/123). 

> As mentioned above, we **do not allow** *dynamic* explicit instantiation. If an expression `e` has type `dynamic` (or `Never` or `Function` ), then <code>e<int></code> is a **compile-time error**. It's not possible to do implicit instantiation without knowing the member signature to some extent, and we also don't allow explicit instantiation. _(Possible alternative: Allow it, and handle it all at run-time, including any errors from having the wrong number or types of arguments, or just not existing at all. We won't do this for now.)_

Author: I do need to escape the `<`. Sadly my editor renders the `<int>`, with it being in a different color as my only hint.

* If *e* denotes a generic top-level, static or local function declaration (again *e* is an identifier or qualified identifier), that declaration must be a generic function declaration, and then <code>*e*\<*typeArgs*></code> performs an explicitly instantiated function tear-off, which works just like the current implicitly instantiated function tear-off except that the types are provided instead of inferred.
* If *e* denotes a generic instance method (*e* has the form <code>*r*.*name*</code> and *r* has a static type for which *name* is a generic interface method), then <code>*e*\<*typeArgs*></code> performs an explicitly instantiated method tear-off, which works just like the current implicitly instantiated method tear-off except that the types are provided instead of inferred.

> * If *e* has a static type which is a generic callable object type (a non-function type with a generic method named `call`), then <code>*e*\<*typeArgs*></code> is equivalent to the instantiated method-tear off <code>*e*\.call<*typeArgs*></code>.

Reviewer 2: OK, so we're basically accepting #1827 and doing it here.

Author: Yes.

* Otherwise, if *e* has a static type which is a generic function type, then <code>*e*\<*typeArgs*></code> is equivalent to the instantiated method-tear off <code>*e*\.call<*typeArgs*></code>.
* Otherwise the expression is a compile-time error. 

>   * This includes *e* having the static type `dynamic` or `Function`. We do not support implicit or explicit instantiation of functions where we do not know the number and bounds of the type parameters at compile-time.

Reviewer 2: 👍 , we promised that to Martin, too!

The `x.a<b,c>` can be an explicitly instantiated generic function tear-off or an explicitly instantiated type literal named using a prefix, which is new. While neither type objects nor functions declare `operator-` or `operator[]`, such could be added using extension methods.

We will disambiguate such situations *heuristically* based on the token following the `>`. In the existing ambiguity we treat `(` as a sign that it's a generic invocation. We extend the number of tokens which, when following a potential type argument list, makes us choose to parse the previous tokens as that type argument list. 

There is a number of tokens which very consistently *end* an expression, and we include all those:

> `)`, `}`, `]`, `;`, `:`, `,`

The we include tokens which we *predict* will continue a generic instantiation:

>  `(`  `.`  `==`  `!=` 

The first six are tokens which cannot possibly start an expression, and therefore cannot occur after a greater-than infix operator. The last four tokens can continue an expression, and of those only `(` can also start an expression, and we already decided how to disambiguate that).

> There are many other tokens which *currently* cannot continue an expression (and therefore cannot validly follow a type argument list) or which cannot *start* an expression (and therefore cannot validly follow a greater-than operator), but in the service of keeping our future options open, we do not choose to depend those at the current time. For example we omit most infix operators, even though they *currently* cannot start a new expression, and therefore cannot follow a `>` infix operator. This leaves us open to allowing some of those operators as prefix operators in the future, like we currently allow `-`.

Reviewer: I'm not sure what "we do not choose to depend those at the current time" means.  Did you mean "we do not chose to include those at the current time"?

Also, at the end of the paragraph, "... like we currently alllow `-`." feels incomplete.  IMHO it would be clearer to say "... like we currently allow with `-`."

Author: I probably meant "to depend on" as in "depending on those not being possible expression starters". I'll expand it.

Reviewer: I can live with the updated text, but IMHO it represents the situation a little inaccurately, because it seems to imply that if we included, say `+`, as a continuation token, we would necessarily be precluded from introducing new syntax in a future release of the language that allowed `+` to follow a `>` operator.  The situation is not really that bad.  The issue is simply that if we included `+` as a continuation token, then code like `f(a < b, c > + d)` would be accepted by the parser today as equivalent to `f((a<b, c>) + d)`, and then in the future if we introduced prefix `+`, the parser would continue to treat `f(a < b, c > + d)` as `f((a<b, c>) + d)`, rather than `f((a < b), (c > +d))`, possibly confusing people in some rare circumstances.

I'm not arguing with the decision, just saying that the rationale for it is not as dire as this paragraph makes it seem.

The `x.a<b,c>` can be an explicitly instantiated generic function tear-off or an explicitly instantiated type literal named using a prefix, which is new. While neither type objects nor functions declare `operator-` or `operator[]`, such could be added using extension methods.

> We will disambiguate such situations *heuristically* based on the token following the `>`. In the existing ambiguity we treat `(` as a sign that it's a generic invocation. We extend the number of tokens which, when following a potential type argument list, makes us choose to parse the previous tokens as that type argument list. 

Reviewer 2: ```suggestion
We will disambiguate such situations *heuristically* based on the token following the `>` that matches the `<` that follows the `<primary>`. In the existing ambiguity we treat `(` as a sign that it's a generic invocation. We extend the number of tokens which, when following a potential type argument list, makes us choose to parse the previous tokens as that type argument list. 
```

Author: It's not necessarily following a primary, it can be a primary or a selector. Rephrased to still talk about the "\> matching the \<"

}
class D extends C with M {
  void method() {
    var f4 = super.inst<int>; // works like (int $) => super.inst<int>($)    
    var f4TypeName = super.inst<int>.runtimeType.toString();  
  }
}
void main() {
  // Type literals.
  var t1 = List<int>; // Type object for `List<int>`.
  var t2 = ListList<int>; // Type object for `List<List<int>>`.

>   // Instantiated function tear-offs.  

Reviewer: Personally I perfer the old version of this code block (with the blank lines)

Author: Done.

  var f5 = d.minst<int>; // int Function(int), works like (int $) => c.minst<int>($);
  var f6 = d.einst<int>; // int Function(int), works like (int $) => Ext(c).einst<int>($); 
  var typeName = List<int>.toString();
  var functionTypeName = local<int>.runtimeType.toString();
}
```

We allow instantiating non-tear-off function values and callable objects.

```dart
T func<T>(T value) => value;
var funcValue = func;
int Function(int) f = funcValue; // Implicitly instantiated.
var f = funcValue<int>; // Explicitly instantiated.

class Id {
  T call<T>(T value) => value;
}
int Function(int) f = Id(); // Implicit instantiation of `Id().call`.
var g = Id()<int>; // Equivalent to `Id().call<int>`
```

Such value instantiation uses the static type of the function for the type arguments, but forwards the actual arguments to the instantiated function value.

That makes a type instantiation expression of the form <code>*e*\<*typeArgs*></code> only allowed if *e* denotes

* a generic type declaration (class, mixin, type alias) and then the result is an instantiated type literal.
* a generic function declaration (top-level, static or local),

> * a generic instance method of a known interface type, or

Reviewer 2: ```suggestion
* an instance method whose static type is a generic function type, or
```

* any other expression which evaluates to a generic function, in which case the result is a non-generic function value. 

When followed by an argument list, we do instantiated invocation (of the unnamed constructor for types) instead of instantiation.

#### Grammar production differences

Our grammar is not as orthogonal as one could hope, mainly for historical reasons.

If *C* denotes a generic class with a constructor `name` and a static generic method `staticMethod` then:

* <code>*C*.name</code> is valid and denotes a generic function.
* <code>*C*\<int>.name</code> is valid and denotes a non-generic function.
* <code>*C*.name\<int></code> is invalid (constructors are not generic).
* <code>*C*.staticMethod</code> is valid and denotes a generic function.
* <code>*C*\<int>.staticMethod</code> is invalid. We do not allow static members through instantiated types.
* <code>*C*\.staticMethod\<int></code> is valid and denotes a non-generic function.
* <code>*C*?.name</code> is invalid, we do not allow that syntax to denote a constructor tear-off.
* <code>*C*?.staticMethod</code> is valid and denotes a generic function.

> * <code>*C*..name</code> is invalid. Cascades on type literals act on the `Type` value instead.

Reviewer 2: ```suggestion
* <code>*C*..name</code> is invalid unless `name` is a member of `Type`. Cascades on type literals act on the `Type` value.
```

Author: Done.
